### PR TITLE
Here is my code review.

### PR DESCRIPTION
**Pull request by palmertt-uc**
**Analysis of Program**

- This mobile application fetches data from an API endpoint that is essentially a coronavirus monitor. The application itself displays this data in a well done graph and a recyclerview which displays the total cases, recovered cases, and total deaths for most countries. From the previous version of the application, the recyclerview was moved to its own separate tab on the bottom navigation bar, where you can find a specific country based off of a drop down list. There is also a news tab on the navigation bar that uses a news API and a recyclerview to display coronavirus related news articles, this also has search functionality.

**Program available on UC Github on time?**

- Yes.

**Program documented/commented well enough for you to understand?**

- Yes. The program provides excellent documentation and comments that very briefly tells what each class, variable and methods do. Almost all of the variable names and methods have meaningful names to the point that you can tell what each variable and method is used for. All of the classes have proper names and the code was very clean with the right amount of space between various methods and variables. On top of this, variables that were similar were grouped together. All of this made the code behind the program very easy to understand.

**Program compile?**

- Yes. The program compiled with no warnings or errors on startup, though the initial home screen did not populate till going to a different screen.

**Rational behind changes**

- Truthfully, it was difficult to find pieces of code that did not align with Kotlin and Android best practices. Therefore, most of my code was more or less cleaning. I changed a few variable names to be more descriptive/have more meaning. I did convert a standard function into a lambda, which will allow for less code. Along with this, I also removed some redundant code and modified various XML files to have proper text resources and content descriptions. Lastly, I added some spacing and Log statements. These changes follow the various Android and Kotlin guidelines (such as accessibility guidelines) and should allow for easier to follow/read code.

**Three technical concepts I learned**

1. How to use multiple APIs in a mobile application.
2. How to use a spinner to select specific items.
3. How to setup a bottom navigation bar. 

**Three technical questions for the group**

1. Are UI elements following Android accessibility guidelines?
2. Do variables have meaningful names?
3. Are UI elements being loaded properly? (See "Program Compile" above)

### CODE REVIEW

File: app/src/main/java/edu/uc/groupproject/covid19tracker/RetrofitClientInstance.kt [hunk 1]
```diff
@@ -19,15 +19,13 @@ object RetrofitClientInstance {
...
-                    }
-                })
-                retrofit = retrofit2.Retrofit.Builder()
+                httpClient.addInterceptor { chain ->
+                    val request =
+                        chain.request().newBuilder().addHeader("x-rapidapi-host", "coronavirus-monitor.p.rapidapi.com")
+                            .addHeader("x-rapidapi-key", "12a8dba6admshd5f767ad7c36e5bp17cb05jsn3c9cee783e36").build()
+                    chain.proceed(request)
+                }
+                retrofit = Retrofit.Builder()
```
Comment (palmertt-uc) [inline app/src/main/java/edu/uc/groupproject/covid19tracker/RetrofitClientInstance.kt:28]: A lambda here makes the code more readable. https://kotlinlang.org/docs/reference/lambdas.html

File: app/src/main/java/edu/uc/groupproject/covid19tracker/service/CasesByCountryDataProvider.kt [hunk 2]
```diff
@@ -25,7 +26,9 @@ class CasesByCountryDataProvider {
         val data = ArrayList<String>()
 
         client.newCall(request).enqueue(object : Callback {
-            override fun onFailure(call: Call, e: IOException) {}
+            override fun onFailure(call: Call, e: IOException) {
+                Log.d("onFailure", "onFailure from getCasesByCountryData called : ${e.printStackTrace()}")
+            }
```
Comment (palmertt-uc) [inline app/src/main/java/edu/uc/groupproject/covid19tracker/service/CasesByCountryDataProvider.kt:31]: This comment is for all Log.d() statements. These may make it easier to debug in the future if any errors occur. https://developer.android.com/reference/kotlin/android/util/Log

File: app/src/main/java/edu/uc/groupproject/covid19tracker/service/NewsDataProvider.kt [hunk 3]
```diff
@@ -1,30 +1,35 @@
...
 import org.json.JSONObject
 import java.io.IOException
 
 class NewsDataProvider {
 
     fun getNewsData(dataType: String): ArrayList<String>?  {
-         var client = OkHttpClient()
-         var request:Request = Request.Builder()
+         val client = OkHttpClient()
+         val request:Request = Request.Builder()
```
Comment (palmertt-uc) [inline app/src/main/java/edu/uc/groupproject/covid19tracker/service/NewsDataProvider.kt]: Declaring variables: https://kotlinlang.org/docs/tutorials/kotlin-for-py/declaring-variables.html

File: app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt [hunk 4]
```diff
@@ -38,9 +37,9 @@ class ByCountryFragment : Fragment() {
          * Declare field and other necessary variables
          */
         val view = inflater.inflate(R.layout.fragment_by_country, container, false)
-        val countryListView: ListView = view.findViewById(R.id.country_list_view) as ListView
-        val countrySpinner: Spinner = view.findViewById(R.id.country_spinner) as Spinner
-        val refreshView: SwipeRefreshLayout = view.findViewById(R.id.by_country_refresh_view) as SwipeRefreshLayout
+        val countryListView: ListView = view.findViewById(R.id.country_list_view)
+        val countrySpinner: Spinner = view.findViewById(R.id.country_spinner)
+        val refreshView: SwipeRefreshLayout = view.findViewById(R.id.by_country_refresh_view)
```
Comment (palmertt-uc) [inline app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt:42]: Removed some redundancy https://kotlinlang.org/docs/reference/coding-conventions.html

File: app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt [hunk 5]
```diff
@@ -67,14 +66,14 @@ class ByCountryFragment : Fragment() {
             /**
              * Declare custom array adapter and set list view
              */
-            val arrAdapter = ItemAdapter(view.context, android.R.layout.simple_list_item_1, casesListViewItems)
-            countryListView.adapter = arrAdapter
+            val casesListViewAdapter = ItemAdapter(view.context, casesListViewItems)
+            countryListView.adapter = casesListViewAdapter
```
Comment (palmertt-uc) [inline app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt:70]: Changed variable names to be more "descriptive" https://kotlinlang.org/docs/reference/coding-conventions.html

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -2,4 +2,18 @@
...
+    <string name="main_fragment_zero">0</string>
+    <string name="main_fragment_local_cases">Local Cases</string>
+    <string name="main_fragment_Deaths">Deaths</string>
+    <string name="main_fragment_recovered">Recovered</string>
+    <string name="news_search_hint">Search</string>
+    <string name="news_fragment_news">News</string>
+    <string name="country_fragment_country_cases">Country Cases</string>
+    <string name="menu_news">News</string>
+    <string name="menu_cases_by_country">Cases by Country</string>
+    <string name="menu_home">Home</string>
```
Comment (palmertt-uc) [inline app/src/main/res/values/strings.xml:18]: https://developer.android.com/guide/topics/resources/string-resource

File: app/src/main/res/layout/news_list_view_layout.xml [hunk 7]
```diff
@@ -10,13 +11,15 @@
...
+        android:orientation="horizontal"
+        tools:ignore="UselessParent">
 
         <ImageView
             android:layout_width="100dp"
             android:layout_height="100dp"
             android:id="@+id/news_image"
-            android:layout_marginRight="10dp"/>
+            android:layout_marginEnd="10dp"
+            android:contentDescription="@string/news_image" />
```
Comment (palmertt-uc) [inline app/src/main/res/layout/news_list_view_layout.xml:22]: https://developer.android.com/guide/topics/ui/accessibility/apps

File: app/src/main/res/layout/fragment_news.xml [hunk 8]
```diff
@@ -15,9 +15,9 @@
         <TextView
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:text="News"
+            android:text="@string/news_fragment_news"
             android:textStyle="bold"
-            android:textSize="20dp"
+            android:textSize="20sp"
```
Comment (palmertt-uc) [inline app/src/main/res/layout/fragment_news.xml:20]: Comment for all dp to sp changes: Use sp when defining text size, it resizes based on user's preferred text size. 
https://developer.android.com/training/multiscreen/screendensities

File: app/src/main/java/edu/uc/groupproject/covid19tracker/RetrofitClientInstance.kt (hunk 1 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/groupproject/covid19tracker/RetrofitClientInstance.kt:28]: Hmmm, good catch here.

File: app/src/main/java/edu/uc/groupproject/covid19tracker/service/CasesByCountryDataProvider.kt (hunk 2 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/groupproject/covid19tracker/service/CasesByCountryDataProvider.kt:31]: Excellent.  Many people leave the onFailure() empty.  Logging here is crucial.  If something goes wrong, the log statement will give you details!

File: app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt (hunk 4 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt:42]: Yes... with Kotlin synthetics, you don't need to specify type.

File: app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt (hunk 5 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/groupproject/covid19tracker/ui/main/ByCountryFragment.kt:70]: Good.  Avoid variable names like "array", "editText1", "test", "a", etc.  Descriptive variable names mean you can spend less time commenting.

File: app/src/main/res/layout/fragment_by_country.xml [hunk 9]
```diff
@@ -15,9 +15,9 @@
         <TextView
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:text="Country Cases"
+            android:text="@string/country_fragment_country_cases"
```
Comment (discospiff) [inline app/src/main/res/layout/fragment_by_country.xml:18]: strings.xml is the best place for string literals.  You can have a different strings.xml for each language you support, and the Android operating system will automatically choose the correct file, based on the user's preferred language.

**Comment by discospiff** [comment]
Numerous excellent suggestions here, which go a long way in reducing technical debt.  I recommend merging this branch.  

It looks like the team already has merged it.

**Pull request merged**